File: classes/myHouse.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from classes.myBattery import Battery
from classes.mySolarPanel import SolarPanels
from classes.myEliaData import EliaData
from classes.myGridData import GridData
from classes.myClimateData import ClimateData
import cvxpy as cp
import numpy as np
from scipy.optimize import minimize

class House:
    def __init__(self, location='', battery=None, solar_panels=None, injection_price=0.04, price_per_kWh=0.35, grid_data=None, elia_data=None, climate_data=None):
        self.location = location
        self.battery = battery if battery is not None else Battery(max_capacity=10, efficiency=0.95)
        self.solar_panels = solar_panels if solar_panels is not None else SolarPanels()
        self.injection_price = injection_price # (eur/kWh)
        self.price_per_kWh = price_per_kWh # (eur/kWh)
        self.grid_data = grid_data if grid_data is not None else GridData()
        self.elia_data = elia_data if elia_data is not None else EliaData()
        self.climate_data = climate_data if climate_data is not None else ClimateData()

        # Variables
        self.import_energy_history = []
        self.export_energy_history = []
        self.import_cost = 0 # Cost for importing energy in gridData (eur)
        self.export_revenue = 0 # Revenue for exporting energy to the grid in gridData(eur)
        self.energy_cost = 0 # import_cost - export_revenue (eur)

        self.optimal_battery_capacity = 0

    def remaining_energy_history(self):
       return list(np.array(self.import_energy_history) - np.array(self.export_energy_history))

    def battery_management_system(self, remaining_energy, current_time):
        return self.dynamic_battery_management_system(remaining_energy, current_time)
        return self.greedy_battery_management_system(remaining_energy)
    
    def greedy_battery_management_system(self, remaining_energy, current_time):
        released_energy, stored_energy = 0, 0
        if remaining_energy > 0:
            # Discharge battery
            released_energy = self.battery.release_energy(remaining_energy)
        else:
            # Charge battery
            stored_energy = self.battery.store_energy(-remaining_energy)
        return released_energy, stored_energy

    def dynamic_battery_management_system(self, remaining_energy, current_time):
        # Next step: predictive control
        # Use previous day as a reference
        released_energy, stored_energy = 0, 0

        # If surplus -> charge
        if remaining_energy <= 0:
            stored_energy = self.battery.store_energy(-remaining_energy)

        # If deficit and SOC > reserve threshold -> discharge
        elif remaining_energy > 0 and self.battery.SOC > self.battery.reserve_soc(current_time):
            released_energy = self.battery.release_energy(remaining_energy)

        # If deficit and SOC ≤ reserve -> cautious strategy
        elif remaining_energy > 0 and self.battery.SOC <= self.battery.reserve_soc(current_time):
            if remaining_energy > self.battery.dynamic_threshold(self.remaining_energy_history()):
                released_energy = self.battery.release_energy(remaining_energy)
            else:
                released_energy = 0

        else:
            raise ValueError("Unhandled BMS case.")

        assert not (released_energy > 0 and stored_energy > 0), \
            "Battery released and stored energy cannot be positive at the same time."

        return released_energy, stored_energy

    def simulate_household(self):
        df = self.grid_data.df
        time_values = df['datetime'].values
        self.import_energy_history = []
        self.export_energy_history = []
        self.import_cost, self.export_revenue, self.energy_cost = 0, 0, 0
        
        for k in range(len(time_values)):
            tmp_remaining_energy = df['remaining'].values[k] # Energy required = Energy net - Energy solar (kWh)
            
            # Battery management
            remaining_required = 0
            remaining_excess = 0

            released_energy, stored_energy = self.battery_management_system(tmp_remaining_energy, time_values[k])

            if tmp_remaining_energy> 0:
                # Discharge battery
                # released_energy = self.battery.release_energy(tmp_remaining_energy)
                remaining_required = tmp_remaining_energy - released_energy
                if remaining_required < 0:
                    raise ValueError("Released energy from battery exceeds remaining required energy.")
                self.energy_cost += remaining_required*self.price_per_kWh  
                self.import_cost += remaining_required*self.price_per_kWh        
            else:
                # Charge battery
                # stored_energy = self.battery.store_energy(-tmp_remaining_energy)
                remaining_excess = -tmp_remaining_energy - stored_energy
                if remaining_excess < -1e-8:
                    raise ValueError("Stored energy exceeds remaining excess energy.")
                self.energy_cost -= remaining_excess * self.injection_price
                self.export_revenue += remaining_excess * self.injection_price
            
            # Save history
            self.import_energy_history.append(remaining_required)
            self.export_energy_history.append(remaining_excess)
            self.battery.SOC_history.append(self.battery.SOC)

        return self.import_energy_history, self.export_energy_history

    def optimize_battery_capacity(self, max_capacity_kwh=15):
        total_cost_array = []
        annualized_battery_cost_array = []
        no_points = 40
        capacity_array = np.linspace(0, max_capacity_kwh, no_points)

        df = self.grid_data.df
        time_values = df['datetime'].values
        time_span = (time_values[-1] - time_values[0]) / np.timedelta64(1, 'D')  # Time span in days
        
        for idx, capacity in enumerate(capacity_array):
            print(f"Progress: {idx+1}/{no_points} (Battery capacity: {capacity:.2f} kWh)")
            self.battery = Battery(max_capacity=capacity)
            self.simulate_household()
            annualized_energy_cost = (self.energy_cost / time_span)*365
            annualized_battery_cost = self.battery.battery_cost() / self.battery.battery_lifetime
            total_cost = annualized_energy_cost + annualized_battery_cost # total annualized cost
            total_cost_array.append(total_cost)
            annualized_battery_cost_array.append(annualized_battery_cost)           

        savings_list = -(total_cost_array - total_cost_array[0]) # Savings compared to no battery
        optimal_idx = np.argmax(savings_list)
        self.optimal_battery_capacity = capacity_array[optimal_idx]
        self.battery = Battery(max_capacity=self.optimal_battery_capacity)
        self.simulate_household()

        return capacity_array, savings_list, annualized_battery_cost_array

    def annualized_cost_function(self):
        time_span = (self.grid_data.df['datetime'].values[-1] - self.grid_data.df['datetime'].values[0]) / np.timedelta64(1, 'D')  # Time span in days
        self.energy_cost = np.sum(self.import_energy_history) * self.price_per_kWh - np.sum(self.export_energy_history) * self.injection_price
        annualized_energy_cost = (self.energy_cost / time_span)*365
        annualized_battery_cost = self.battery.battery_cost() / self.battery.battery_lifetime
        return annualized_energy_cost + annualized_battery_cost

    def optimize_battery_management(self):
        initial_battery = self.battery
        df = self.grid_data.df

        # Simulate household without battery
        self.battery = Battery(max_capacity=0)
        self.simulate_household()
        total_cost_0 = self.annualized_cost_function()
        self.plot_energy_history()
        
        # Simulate household with greedy BMS
        self.battery_management_system = self.greedy_battery_management_system
        self.battery = initial_battery
        self.simulate_household()
        total_cost_greedy = self.annualized_cost_function()
        self.plot_energy_history()
        
        # Simulate household with dynamic BMS
        self.battery_management_system = self.dynamic_battery_management_system
        self.battery = initial_battery
        self.simulate_household()
        total_cost_dynamic = self.annualized_cost_function()
        self.plot_energy_history()
        
        # Simulate household with optimized battery management using convex optimization
        self.battery = initial_battery # reset to initial battery
        time_vec = (df['datetime'] - df['datetime'].iloc[0]).dt.total_seconds().values / 3600
        dt = np.diff(time_vec).mean()

        P_load = df['remaining'].values  # kW demand
        n = len(P_load)

        # Decision variables
        P_b = cp.Variable(n)        # battery net power (discharge - charge)
        SOC = cp.Variable(n+1)      # state of charge
        P_import = cp.Variable(n, nonneg=True)
        P_export = cp.Variable(n, nonneg=True)
        P_charge = cp.Variable(n, nonneg=True)
        P_discharge = cp.Variable(n, nonneg=True)

        # Parameters
        SOC_init = self.battery.SOC
        SOC_max = 1
        P_max = self.battery.C_rate * self.battery.max_capacity

        # Constraints
        constraints = []
        constraints.append(SOC[0] == SOC_init)

        for t in range(n):
            # SOC update (DCP-compliant)
            constraints.append(SOC[t+1] == SOC[t] + (self.battery.efficiency * P_charge[t] - P_discharge[t] / self.battery.efficiency ) / self.battery.max_capacity)
            constraints.append(SOC[t+1] >= 0)
            constraints.append(SOC[t+1] <= SOC_max)
            constraints.append(P_charge[t] >= 0)
            constraints.append(P_discharge[t] >= 0)
            constraints.append(P_charge[t] <= P_max)
            constraints.append(P_discharge[t] <= P_max)
            # Link battery net power to charge/discharge
            constraints.append(P_b[t] == P_discharge[t] - P_charge[t])
            constraints.append(P_b[t] <= P_max)
            constraints.append(P_b[t] >= -P_max)
            # Import/export constraints
            constraints.append(P_import[t] - P_export[t] == P_load[t] - P_b[t])
            constraints.append(P_import[t] >= 0)
            constraints.append(P_export[t] >= 0)

        # Convex objective
        cost = cp.sum(P_import * self.price_per_kWh - P_export * self.injection_price)

        # Solve
        prob = cp.Problem(cp.Minimize(cost), constraints)
        prob.solve()

        self.import_energy_history = P_import.value
        self.export_energy_history = P_export.value
        self.battery.SOC_history = SOC.value.tolist()

        total_cost_optim = self.annualized_cost_function()
        self.plot_energy_history()
        
        return self.import_energy_history, self.export_energy_history
        

    def plot_energy_history(self):
        df = self.grid_data.df
        time = df['datetime'].values
        
        fig, axes = plt.subplots(1, 2, figsize=(16, 6))
        # Line plot
        axes[0].plot(time, self.import_energy_history, label='Import Energy (kWh)', color='red')
        axes[0].plot(time, self.export_energy_history, label='Export Energy (kWh)', color='green')
        axes[0].set_title(f'Energy History (Battery Capacity: {self.battery.max_capacity:.2f} kWh)')
        axes[0].set_xlabel('Time (hours)')
        axes[0].set_ylabel('Energy (kWh)')
        axes[0].legend()
        axes[0].grid()

        # Add price annotation (energy cost)
        price_text = f"Total Energy Cost: {self.energy_cost:.2f} EUR\nImport Cost: {self.import_cost:.2f} EUR\nExport Revenue: {self.export_revenue:.2f} EUR"
        axes[0].text(0.05, 0.95, price_text, transform=axes[0].transAxes, fontsize=10, verticalalignment='top', bbox=dict(boxstyle="round,pad=0.3", facecolor="lightyellow", alpha=0.5))

        # Histogram of simulated remaining energy
        axes[1].hist(self.remaining_energy_history(), bins=30, color='purple', alpha=0.7)
        axes[1].set_title('Histogram of Remaining Consumption')
        axes[1].set_xlabel('Remaining Consumption (kWh)')
        axes[1].set_ylabel('Frequency')
        axes[1].grid(True)

        plt.tight_layout()
        return fig
    

File: fluvius_calculations_flutter/lib/classes/myHouse.dart
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:fluvius_calculations_flutter/classes/myBattery.dart';
import 'package:fluvius_calculations_flutter/classes/myGridData.dart';
import 'package:intl/intl.dart';

class House extends ChangeNotifier {
  String location = '';
  double injection_price = 0.05; // (€/kWh)
  double price_per_kWh = 0.35; // (€/kWh)
  Battery battery = Battery();
  GridData grid_data = GridData();

  List<double> import_energy_history =
      []; // List to store import energy history
  List<double> export_energy_history =
      []; // List to store export energy history
  double import_cost = 0.0;
  double export_revenue = 0.0;
  double energy_cost = 0.0;
  double optimal_battery_capacity = 0.0;
  List<double> capacity_array = [];
  List<double> savings_list = [];
  List<double> annualized_battery_cost_array = [];

  void updateParameters({
    String? newLocation,
    double? newInjectionPrice,
    double? newPricePerKWh,
    Battery? newBattery,
    GridData? newGridData,
    List<double>? newImportEnergyHistory,
    List<double>? newExportEnergyHistory,
    double? newImportCost,
    double? newExportRevenue,
    double? newEnergyCost,
    double? newOptimalBatteryCapacity,
    List<double>? capacity_array,
    List<double>? savings_list,
    List<double>? annualized_battery_cost_array,
    String? newBase64Image,
  }) {
    if (newLocation != null) {
      location = newLocation;
    }
    if (newInjectionPrice != null) {
      injection_price = newInjectionPrice;
    }
    if (newPricePerKWh != null) {
      price_per_kWh = newPricePerKWh;
    }
    if (newBattery != null) {
      battery = newBattery;
    }
    if (newGridData != null) {
      grid_data = newGridData;
    }
    if (newImportEnergyHistory != null) {
      import_energy_history = newImportEnergyHistory;
    }
    if (newExportEnergyHistory != null) {
      export_energy_history = newExportEnergyHistory;
    }
    if (newImportCost != null) {
      import_cost = newImportCost;
    }
    if (newExportRevenue != null) {
      export_revenue = newExportRevenue;
    }
    if (newEnergyCost != null) {
      energy_cost = newEnergyCost;
    }
    if (newOptimalBatteryCapacity != null) {
      optimal_battery_capacity = newOptimalBatteryCapacity;
    }
    if (capacity_array != null) {
      capacity_array = capacity_array;
    }
    if (savings_list != null) {
      savings_list = savings_list;
    }
    if (annualized_battery_cost_array != null) {
      annualized_battery_cost_array = annualized_battery_cost_array;
    }
    notifyListeners();
  }

  // Battery Management System methods
  Map<String, double> batteryManagementSystem(
    double remainingEnergy,
    DateTime currentTime,
  ) {
    return greedyBatteryManagementSystem(remainingEnergy);
    // return dynamicBatteryManagementSystem(remainingEnergy, currentTime);
  }

  Map<String, double> greedyBatteryManagementSystem(double remainingEnergy) {
    double releasedEnergy = 0;
    double storedEnergy = 0;

    if (remainingEnergy > 0) {
      // Discharge battery
      releasedEnergy = battery.releaseEnergy(remainingEnergy);
    } else {
      // Charge battery
      storedEnergy = battery.storeEnergy(-remainingEnergy);
    }

    return {'released': releasedEnergy, 'stored': storedEnergy};
  }

  Map<String, double> dynamicBatteryManagementSystem(
    double remainingEnergy,
    DateTime currentTime,
  ) {
    double releasedEnergy = 0;
    double storedEnergy = 0;

    // If surplus -> charge
    if (remainingEnergy <= 0) {
      storedEnergy = battery.storeEnergy(-remainingEnergy);
    }
    // If deficit and SOC > reserve threshold -> discharge
    else if (remainingEnergy > 0 &&
        battery.SOC > battery.reserveSoc(currentTime)) {
      releasedEnergy = battery.releaseEnergy(remainingEnergy);
    }
    // If deficit and SOC ≤ reserve -> cautious strategy
    else if (remainingEnergy > 0 &&
        battery.SOC <= battery.reserveSoc(currentTime)) {
      final threshold = battery.dynamicThreshold(remainingEnergyHistory());
      if (remainingEnergy > threshold) {
        releasedEnergy = battery.releaseEnergy(
          remainingEnergy * 0.5,
        ); // Conservative discharge
      }
      // Else: Don't discharge, preserve battery for critical needs
    }

    return {'released': releasedEnergy, 'stored': storedEnergy};
  }

  List<double> remainingEnergyHistory() {
    final List<double> remaining = [];
    for (
      int i = 0;
      i < import_energy_history.length && i < export_energy_history.length;
      i++
    ) {
      remaining.add(import_energy_history[i] - export_energy_history[i]);
    }
    return remaining;
  }

  // Main simulation method
  Map<String, dynamic> simulateHousehold() {
    if (grid_data.processedData.isEmpty) {
      return {
        'success': false,
        'message':
            'No processed data available. Please load and process CSV data first.',
      };
    }

    // Reset simulation data
    import_energy_history.clear();
    export_energy_history.clear();
    battery.SOC_history.clear();
    battery.start_date = grid_data.start_date;
    battery.end_date = grid_data.end_date;
    import_cost = 0;
    export_revenue = 0;
    energy_cost = 0;

    try {
      for (final dataPoint in grid_data.processedData) {
        final remainingEnergy =
            dataPoint.volume_afname_kwh - dataPoint.volume_injectie_kwh;

        // Battery management
        double remainingRequired = 0;
        double remainingExcess = 0;

        final bmsResult = batteryManagementSystem(
          remainingEnergy,
          dataPoint.datetime,
        );
        final releasedEnergy = bmsResult['released'] ?? 0;
        final storedEnergy = bmsResult['stored'] ?? 0;

        if (remainingEnergy > 0) {
          // Energy deficit - need to import from grid
          remainingRequired = remainingEnergy - releasedEnergy;
          if (remainingRequired < 0) {
            remainingRequired = 0; // Can't be negative
          }
        } else {
          // Energy surplus - may export to grid
          remainingExcess = (-remainingEnergy) - storedEnergy;
          if (remainingExcess < 0) remainingExcess = 0; // Can't be negative
        }

        // Save history
        import_energy_history.add(remainingRequired);
        export_energy_history.add(remainingExcess);
        battery.SOC_history.add(battery.SOC);
      }

      // Calculate costs
      _calculateCosts();

      notifyListeners();

      return {
        'success': true,
        'import_energy': import_energy_history,
        'export_energy': export_energy_history,
        'soc_history': battery.SOC_history,
        'import_cost': import_cost,
        'export_revenue': export_revenue,
        'energy_cost': energy_cost,
        'message': 'Simulation completed successfully',
      };
    } catch (e) {
      return {'success': false, 'message': 'Simulation failed: $e'};
    }
  }

  void _calculateCosts() {
    import_cost = import_energy_history.fold(
      0.0,
      (sum, energy) => sum + energy * price_per_kWh,
    );
    export_revenue = export_energy_history.fold(
      0.0,
      (sum, energy) => sum + energy * injection_price,
    );
    energy_cost = import_cost - export_revenue;
  }

  // Cost calculation method
  double annualizedCostFunction() {
    if (grid_data.processedData.isEmpty) return 0;

    final firstTime = grid_data.processedData.first.datetime;
    final lastTime = grid_data.processedData.last.datetime;
    final timeSpanDays = lastTime.difference(firstTime).inDays.toDouble();

    if (timeSpanDays <= 0) return 0;

    final annualizedEnergyCost = (energy_cost / timeSpanDays) * 365;
    final annualizedBatteryCost =
        battery.batteryCost() / battery.battery_lifetime;

    return annualizedEnergyCost + annualizedBatteryCost;
  }

  // Optimization
  Map<String, dynamic> optimizeBatteryCapacity({
    double minCapacity = 0.0,
    double maxCapacity = 20.0,
    double stepSize = 1,
  }) {
    try {
      capacity_array.clear();
      savings_list.clear();
      annualized_battery_cost_array.clear();

      double bestCapacity = minCapacity;
      double lowestCost = double.infinity;

      for (
        double capacity = minCapacity;
        capacity <= maxCapacity;
        capacity += stepSize
      ) {
        battery.max_capacity = capacity;
        simulateHousehold();
        final totalCost = annualizedCostFunction();

        capacity_array.add(capacity);
        annualized_battery_cost_array.add(
          battery.batteryCost() / battery.battery_lifetime,
        );
        savings_list.add(annualizedCostFunction());

        if (totalCost < lowestCost) {
          lowestCost = totalCost;
          bestCapacity = capacity;
        }
      }

      optimal_battery_capacity = bestCapacity;
      notifyListeners();

      return {
        'success': true,
        'optimal_capacity': bestCapacity,
        'lowest_annualized_cost': lowestCost,
        'message': 'Optimization completed successfully',
      };
    } catch (e) {
      return {'success': false, 'message': 'Simulation failed: $e'};
    }
  }

  Map<String, dynamic> toJson() {
    return {
      'location': location,
      'injection_price': injection_price,
      'price_per_kWh': price_per_kWh,
      'battery': battery.toJson(),
      'grid_data': grid_data.toJson(),
    };
  }
}
